File: haskell/TestFramework.hs
module TestFramework where

data Test = Test
    { failMsg :: String
    , runTest :: Bool
    }

testEqual :: (Show a, Eq a) => String -> a -> a -> Test
testEqual msg expected actual = Test elaborateMsg (expected == actual)
    where
    elaborateMsg :: String
    elaborateMsg = msg ++ "\n    Expected: " ++ show expected
        ++ "\n    Actual: " ++ show actual

testMain :: [Test] -> IO ()
testMain = tm 0 0
    where
    tm :: Int -> Int -> [Test] -> IO ()
    tm total pass [] = putStrLn $
        "\nTests run: " ++
        (show pass) ++ " passed / " ++
        (show (total - pass)) ++ " failed / " ++
        (show total) ++ " total. "
    tm total pass (test:rest) = case runTest test of
        True -> tm (total + 1) (pass + 1) rest
        False -> do
            putStrLn $ "FAIL: " ++ (failMsg test)
            tm (total + 1) pass rest

File: python/TestFramework.py
from dataclasses import dataclass
from typing import Callable, List


@dataclass
class Test:
    failMsg: str
    runTest: bool


def testEqual(msg: str, expected, actual) -> Test:
    elaborate = f"{msg}\n    Expected: {expected}\n    Actual: {actual}"
    return Test(elaborate, expected == actual)


def testMain(tests: List[Test]) -> None:
    total = 0
    passed = 0
    for test in tests:
        total += 1
        if test.runTest:
            passed += 1
        else:
            print(f"FAIL: {test.failMsg}")
    print(f"\nTests run: {passed} passed / {total - passed} failed / {total} total. ")
